vello/src/scene.rs

```diff
@@ -334,23 +345,284 @@ impl<'a> DrawGlyphs<'a> {
+                return;
+            }
+            let index = resources.glyph_runs.len();
+            resources.glyph_runs.push(self.run);
+            resources.patches.push(Patch::GlyphRun { index });
+            self.scene
+                .encoding
+                .encode_brush(self.brush, self.brush_alpha);
+            // Glyph run resolve step affects transform and style state in a way
+            // that is opaque to the current encoding.
+            // See <https://github.com/linebender/vello/issues/424>
+            self.scene.encoding.force_next_transform_and_style();
+        }
+    }
+
+    pub fn try_draw_colr(&mut self, glyphs: impl Iterator<Item = Glyph>) {
+        let font_index = self.run.font.index;
+        let font = skrifa::FontRef::from_index(self.run.font.data.as_ref(), font_index).unwrap();
+        let upem: f32 = font.head().map(|h| h.units_per_em()).unwrap().into();
+        let run_transform = self.run.transform.to_kurbo();
+        let scale = Affine::scale_non_uniform(
+            (self.run.font_size / upem).into(),
+            (-self.run.font_size / upem).into(),
+        );
+        let colour_collection = font.color_glyphs();
+        for glyph in glyphs {
+            let colour = colour_collection
+                .get(GlyphId::new(glyph.id.try_into().unwrap()))
+                .expect("Cannot render non-emoji in emoji font");
+            let coords = &self.scene.encoding.resources.normalized_coords
+                [self.run.normalized_coords.clone()]
+            .to_vec();
+            let transform = run_transform
+                * Affine::translate(Vec2::new(glyph.x.into(), glyph.y.into()))
+                * scale
+                * self
+                    .run
+                    .glyph_transform
+                    .unwrap_or(Transform::IDENTITY)
+                    .to_kurbo();
+            colour
+                .paint(
+                    LocationRef::new(coords),
+                    &mut DrawColorGlyphs {
+                        scene: &mut self.scene,
+                        cpal: &font.cpal().unwrap(),
+                        outlines: &font.outline_glyphs(),
+                        transform_stack: vec![Transform::from_kurbo(&transform)],
+                    },
+                )
+                .unwrap();
+        }
+    }
+}
+
+struct DrawColorGlyphs<'a> {
+    scene: &'a mut Scene,
+    transform_stack: Vec<Transform>,
+    cpal: &'a Cpal<'a>,
+    outlines: &'a OutlineGlyphCollection<'a>,
+}
+
+impl ColorPainter for DrawColorGlyphs<'_> {
+    fn push_transform(&mut self, transform: skrifa::color::Transform) {
+        // eprintln!("Pushing a transform");
+        let transform = conv_skrifa_transform(transform);
+        let prior_transform = self.last_transform();
+        self.transform_stack.push(prior_transform * transform);
+    }
+
+    fn pop_transform(&mut self) {
+        self.transform_stack.pop();
+    }
+
+    fn push_clip_glyph(&mut self, _: GlyphId) {
```

Comment:
This really does not match the Vello imaging model - I'm inclined to leave glyphs which require this as unsupported for the moment

Reply:
This is just a `push_layer` with the glyph outline as the path and the `Mix::Clip` blending mode.

Reply:
(and the current transform)

Reply:
How so? Clips and layers share the blend stack in vello. We shouldn’t need to track anything but transforms.

Reply:
There's no way to implement "fill the entire current clip area" from the scene API, as far as I know
Nor is there "add a layer atop the entire current clip area"

Which means we need to store the innermost clip area at all times, which as far as I can requires a new stack of the outlines/bezpaths. Maybe I'm missing some bit of API surface? 

And if we get a:
```
push_clip
push_layer
pop_clip
pop_layer
```
interface, we need even more messiness.
Plus we have to ignore a `push_layer` before anything else, because that's meaningless, but then we have to ignore the corresponding `pop_layer`. It seems to be a massive headache which I would rather not deal with if we can get away with it

Reply:
There are definitely ways to produce degenerate scenes but the flattened representation is generated from a tree so clips/layers will always be properly ordered and popped assuming there’s not a parsing error.

This work looks great to me so far and I’d be happy to pick it up if you’d like to merge and move on.